### PR TITLE
[VL] Enable spark rand function

### PR DESCRIPTION
**Pull request by gaoyangxiaozhu**
## What changes were proposed in this pull request?

velox already support spark rand function with `seed + partitionId` as actual seed, checking this https://github.com/facebookincubator/velox/blob/a0222abbafd78678a3a3fed550d1765230374c84/velox/functions/sparksql/Rand.h#L31C1-L33C41 , so we can enable it right now.

this is the dependency PR for `SampleExec` support whichi need rand end to end in velox ready https://github.com/apache/incubator-gluten/issues/5315

## How was this patch tested?

manually test run

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala [hunk 1]
```diff
@@ -454,9 +454,9 @@ object ExpressionConverter extends SQLConfHelper with Logging {
           m
         )
       case rand: Rand =>
-        BackendsApiManager.getSparkPlanExecApiInstance.genRandTransformer(
+        GenericExpressionTransformer(
           substraitExprName,
-          replaceWithExpressionTransformerInternal(rand.child, attributeSeq, expressionsMap),
+          Seq(replaceWithExpressionTransformerInternal(rand.child, attributeSeq, expressionsMap)),
```
Comment (rui-mo) [inline gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala:459]: I wonder if we still need this specialization as the default branch could produce a GenericExpressionTransformer.

**Comment by gaoyangxiaozhu** [comment]
@zhouyuan and @rui-mo looks the online CI fail due to still use vinalia spark rand not velox rand , but the `random.sql` and `group-by.sql` tests work locally, any clues or suggestions to debug for this ?

image

**Comment by rui-mo** [comment]
> looks the online CI fail due to still use vinalia spark rand not offload to use velox rand

@gaoyangxiaozhu One possibility I can think of is the disturbing from Presto rand function as we register both. Since its signature differs from Spark's, validation fails and the rand function fallbacks to vanilla Spark.

**Comment by gaoyangxiaozhu** [comment]
> @zhouyuan and @rui-mo looks the online CI fail due to still use vinalia spark rand not offload to use velox rand , but in locally test it works, the `random.sql` and `group-by.sql` tests work as expected when I run locally, any clues or suggestions to debug for this ?
> 
> image

ping again

**Comment by rui-mo** [comment]
@gaoyangxiaozhu Is it due to the issue I mentioned in https://github.com/apache/incubator-gluten/pull/5829#issuecomment-2126051621? Thanks.

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu Is it due to the issue I mentioned in [#5829 (comment)](https://github.com/apache/incubator-gluten/pull/5829#issuecomment-2126051621)? Thanks.

hey @rui-mo , will check, but if it is the issue why local test passed and also do you know however to verify quickly ?

**Comment by rui-mo** [comment]
@gaoyangxiaozhu Just notice this PR seems to be similar as https://github.com/apache/incubator-gluten/pull/4879, where we also notice this issue and are fixing with https://github.com/apache/incubator-gluten/pull/4879/files#diff-2b284204ae42824909be5636f11618b207a55688757b510d3b3890c44065e41dR397-R403.

**Comment by rui-mo** [comment]
> do you know how to verify quickly

@gaoyangxiaozhu We can differentiate the names used by Presto and Spark and verify if CI passes.

**Comment by gaoyangxiaozhu** [comment]
> @gaoyangxiaozhu Just notice this PR seems to be similar as #4879, where we also notice this issue and are fixing with https://github.com/apache/incubator-gluten/pull/4879/files#diff-2b284204ae42824909be5636f11618b207a55688757b510d3b3890c44065e41dR397-R403.

got,  so do we have a PR in velox side to update registering `rand` with name `spark_rand` available ?

**Comment by rui-mo** [comment]
@gaoyangxiaozhu I think the registration is covered by https://github.com/apache/incubator-gluten/pull/4879/files#diff-36df5e0d154adae284cafe3d831d6d6fe7096983b4975c600a35dc1ce2b8cfa4R47-R50.

File: gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala (hunk 1 above)
Comment (gaoyangxiaozhu) [reply gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala]: good suggestion!

**Comment by gaoyangxiaozhu** [comment]
thank you @rui-mo  now the PR passed CI, please help review the change when you have time. thank you again!

File: cpp/velox/operators/functions/RegistrationAllFunctions.cc [hunk 2]
```diff
@@ -43,6 +44,9 @@ void registerFunctionOverwrite() {
   velox::registerFunction<RoundFunction, int64_t, int64_t, int32_t>({"round"});
   velox::registerFunction<RoundFunction, double, double, int32_t>({"round"});
   velox::registerFunction<RoundFunction, float, float, int32_t>({"round"});
+  // TODO: the below rand function registry can be removed after presto function registry is removed.
+  velox::registerFunction<velox::functions::sparksql::RandFunction, double, velox::Constant<int32_t>>({"spark_rand"});
+  velox::registerFunction<velox::functions::sparksql::RandFunction, double, velox::Constant<int64_t>>({"spark_rand"});
```
Comment (rui-mo) [inline cpp/velox/operators/functions/RegistrationAllFunctions.cc:49]: cc: @PHILO-HE

**Review by rui-mo — COMMENTED** [review]
Thanks! Looks good overall.

File: gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala [hunk 3]
```diff
@@ -1132,6 +1132,7 @@ class VeloxTestSettings extends BackendTestSettings {
     .exclude("SPARK-41017: filter pushdown with nondeterministic predicates")
   enableSuite[GlutenFileScanSuite]
   enableSuite[GlutenGeneratorFunctionSuite]
+    .exclude("SPARK-45171: Handle evaluated nondeterministic expression")
```
Comment (rui-mo) [inline gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala:1135]: Why do we add this test but also exclude it?

Comment (gaoyangxiaozhu) [reply gluten-ut/spark35/src/test/scala/org/apache/gluten/utils/velox/VeloxTestSettings.scala:1135]: the excluded test is spark side test not new added one, the new added one has `Gluten -` prefix

**Review by rui-mo — APPROVED** [review]

**Comment by gaoyangxiaozhu** [comment]
@zhli1142015 help merge

**Pull request merged**